### DISCUSSION TITLE
请问使用 scripts/test.lua 测试包的时候，如何指定依赖库的版本？

### DISCUSSION THREAD

**Discussion by ivanallen** [xmake-io/xmake#4079] [Q&A]
在测试自制包的场景下，如何指定依赖库的版本？
另外在构建预编译仓库的时候，使用的 scripts/build.lua ，也不知道如何去指定。

**Comment by waruqi** [comment]
> 在测试自制包的场景下，如何指定依赖库的版本？

add_deps("xxx 1.0")


> 另外在构建预编译仓库的时候，使用的 scripts/build.lua ，也不知道如何去指定。

指定不了，ci 上自动跑的

**Reply by ivanallen** [reply]
我这边自建了一个预编译仓库。我发现预编译仓库构建时，总是依赖最新的（因为包里没有显式指定）的子依赖。
像 protobuf 这种不同版本无法兼容的情况，预编译的包就可能会有问题。因为使用方可能会使用另一个不同的 protobuf 版本。

**Reply by waruqi** [reply]
这个目前没办法，要是每个依赖的版本不同，都要重新预编译，这个排列组合就多了，ci ，存储资源有限。。只能预编译部分常用组合。。

如果有问题，可以暂时禁用预编译，走源码构建。。{build = true}

**Reply by ivanallen** [reply]
我现在是打算自己做一个常用组合出来，但是似乎无法在 build.txt 中配置依赖库的版本

**Reply by waruqi** [reply]
你可以自己改下脚本

**Reply by ivanallen** [reply]
我看 build.lua 脚本里调用的实际上也是
```lua
local argv = {"lua", "private.xrepo", "install", "-yvD", "--shallow", "--build", "--linkjobs=2", "-p", opt.plat, "-a", opt.arch, "-k", opt.kind}
if opt.configs then
    table.insert(argv, "-f")
    table.insert(argv, opt.configs)
end
```
传递依赖库的版本，是使用 configs 进行传递吗？比如我 install grpc，指定 grpc 依赖的protobuf 版本，应该如何传参？
```c++
# 可能想要个类似的命令
xrepo install -yvD --shallow --build grpc  --deps "protobuf 3.14.0" --deps "zlib 1.0.1"
```

**Reply by waruqi** [reply]
xrepo install /xxxx/packages.lua

可以指定一个 lua 脚本，里面走 add_requires/add_requireconfs 去配置包依赖安装。。但是要自己改脚本

**Reply by ivanallen** [reply]
好吧，看起来有点困难，还是不知道怎么弄。。。

**Reply by star-hengxing** [reply]
用 on_donwload 可行？
